### ISSUE TITLE
How to realize LightGCN model in the paper?

### ISSUE THREAD

**Issue by chenjiyan2001** [tanatosuu/GDE#4]
Hi, I'm very interested in your resaerch, I tried to implement the LightGCN model according to the paper and my own understanding, but there is a big gap with the results reported in the paper. I wonder if you can provide the code corresponding to the LightGCN model? The following is my implementation:
```python
class LightGCN(nn.Module):
	def __init__(self, user_size, item_size, n_layers, beta=4.0, feature_type='both', drop_out=0.2, latent_size=64, reg=0.01):
		super(LightGCN, self).__init__()
		self.user_embed=torch.nn.Embedding(user_size,latent_size)
		self.item_embed=torch.nn.Embedding(item_size,latent_size)

		nn.init.xavier_normal_(self.user_embed.weight)
		nn.init.xavier_normal_(self.item_embed.weight)
		self.n_layers=n_layers
		self.beta=beta
		self.reg=reg
		self.drop_out=drop_out
		if drop_out!=0:
			self.m=torch.nn.Dropout(drop_out)

		if feature_type=='smoothed':
			user_filter=torch.Tensor(np.load(r'./'+dataset+ r'/'+dataset+'_smooth_user_values.npy')).cuda()
			item_filter=torch.Tensor(np.load(r'./'+dataset+ r'/'+dataset+'_smooth_item_values.npy')).cuda()

			user_vector=torch.Tensor(np.load(r'./'+dataset+ r'/'+dataset+'_smooth_user_features.npy')).cuda()
			item_vector=torch.Tensor(np.load(r'./'+dataset+ r'/'+dataset+'_smooth_item_features.npy')).cuda()


		elif feature_type=='both':

			user_filter=torch.cat([torch.Tensor(np.load(r'./'+dataset+ r'/'+dataset+'_smooth_user_values.npy')).cuda()\
				,torch.Tensor(np.load(r'./'+dataset+ r'/'+dataset+'_rough_user_values.npy')).cuda()])

			item_filter=torch.cat([torch.Tensor(np.load(r'./'+dataset+ r'/'+dataset+'_smooth_item_values.npy')).cuda()\
				,torch.Tensor(np.load(r'./'+dataset+ r'/'+dataset+'_rough_item_values.npy')).cuda()])


			user_vector=torch.cat([torch.Tensor(np.load(r'./'+dataset+ r'/'+dataset+'_smooth_user_features.npy')).cuda(),\
				torch.Tensor(np.load(r'./'+dataset+ r'/'+dataset+'_rough_user_features.npy')).cuda()],1)


			item_vector=torch.cat([torch.Tensor(np.load(r'./'+dataset+ r'/'+dataset+'_smooth_item_features.npy')).cuda(),\
				torch.Tensor(np.load(r'./'+dataset+ r'/'+dataset+'_rough_item_features.npy')).cuda()],1)


		else:
			print('error')
			exit()

		self.L_u=(user_vector*user_filter).mm(user_vector.t())
		self.L_i=(item_vector*item_filter).mm(item_vector.t())


		del user_vector,item_vector,user_filter, item_filter
		gc.collect()
		torch.cuda.empty_cache()

	
	# epochs: 180 nDCG@20: 0.381 Recall@20: 0.357
	def forward(self, user, pos_item, nega_item, loss_type='bpr'):
		user_emb, item_emb, nega_emb = self.user_embed.weight, self.item_embed.weight, self.item_embed.weight
		user_embs, item_embs, nega_embs = [user_emb], [item_emb], [nega_emb]
		for layer in range(self.n_layers):
			if self.drop_out==0:
				user_emb,item_emb,nega_emb=self.L_u.mm(user_emb),self.L_i.mm(item_emb),self.L_i.mm(nega_emb)

			else:
				user_emb,item_emb,nega_emb=(self.m(self.L_u)*(1-self.drop_out)).mm(user_emb),(self.m(self.L_i)*(1-self.drop_out)).mm(item_emb),\
				(self.m(self.L_i)*(1-self.drop_out)).mm(nega_emb)

			user_embs.append(user_emb), item_embs.append(item_emb), nega_embs.append(nega_emb)

		user_embs, item_embs, nega_embs = torch.stack(user_embs, dim=1), torch.stack(item_embs, dim=1), torch.stack(nega_embs, dim=1)
		final_user,final_pos,final_nega=torch.mean(user_embs, dim=1)[u],torch.mean(item_embs, dim=1)[i],torch.mean(nega_embs, dim=1)[nega]
        
		if loss_type=='adaptive':

			res_nega=(final_user*final_nega).sum(1)
			nega_weight=(1-(1-res_nega.sigmoid().clamp(max=0.99)).log10()).detach()
			out=((final_user*final_pos).sum(1)-nega_weight*res_nega).sigmoid()

		else:	
			out=((final_user*final_pos).sum(1)-(final_user*final_nega).sum(1)).sigmoid()

		reg_term=self.reg*(final_user**2+final_pos**2+final_nega**2).sum()
		return (-torch.log(out).sum()+reg_term)/batch_size
```

**Comment by tanatosuu** [comment]
It seems you directly use self.L_u  and self.L_i for propagation, but they correspond to the matrices in parenthesis in Eq (11, 12), which already captures higher-order neighborhood from any-hops. The original LightGCN uses the normalized adjacency matrix of a simple graph for propagation (Eq (6) in LightGCN paper), and I have two ways for implementation of LightGCN (I didn't read their codes on Github).

The first one is straightforward and simplest. Take a look at Eq (8) in my paper, LightGCN basically constists of two parts: **\sum_L^k** summing the power of Adjacency matrix and **E**  for user/item embedding, we can separate the two parts, calculate the sum of the power of adjacency matrix **\sum_L^k**, cosidering them as one matrix (e.g., sum_L). Then during training, the model is simplifed as sum_L.mm(E), which is easy to implement and runs vert fast (you can even update node-wisely, which the following way cannot). The problem of this way is space complexity is high especially when datasets are large.


The second one is based on the implementation of other GCN-based recommedation methods (probably close to the offical LightGCN code). Basically the adjacency matrix is sparse with only R and R^T, so we only need this R for propagation. The problems is the propgation rule has to be repeated via each update. The core in my code is like:

```
#I didn't use the standard way to implement with pytorch
def create_embeddings():

	final_user_embedding,final_item_embedding=user_embeddings,item_embeddings
	previous_user,previous_item=user_embeddings,item_embeddings

	for i in range(layers):

               #sparse multiplcation can speed-up calculation
		now_user=torch.sparse.mm(R,previous_item)
		now_item=torch.sparse.mm(R.t(),previous_user)

               #accumulate
		final_user_embedding=final_user_embedding+now_user
		final_item_embedding=final_item_embedding+now_item
		

		previous_user,previous_item=now_user,now_item


	return final_user_embedding,final_item_embedding
```

For reference: https://github.com/newlei/LR-GCCF

**Comment by tanatosuu** [comment]
For the first way, the core code is very simple:

```
#Lets say you alreday loaded sum_L. u, p, nega for user, pos_item, nega_item sets.

user_embedding,pos_embedding,nega_embedding=sum_L[u].mm(E),sum_L[user+p].mm(E),sum_L[user+nega].mm(E)

out=((user_embedding*pos_embedding).sum(1)-(user_embedding*nega_embedding).sum(1)).sigmoid()
```

**Comment by chenjiyan2001** [comment]
Thank you very much for your detailed answers! It helped me a lot

**Comment by tanatosuu** [comment]
Glad it helped. Feel free to drop any questions:)
